include/dca/linalg/util/magma_queue.hpp

```diff
@@ -13,39 +13,66 @@
 #define DCA_LINALG_UTIL_MAGMA_QUEUE_HPP
 #ifdef DCA_HAVE_CUDA
 
+#include <cublas_v2.h>
 #include <cuda.h>
-#include <magma.h>
+#include <cusparse_v2.h>
+#include <magma_v2.h>
+
+#include "dca/linalg/util/cuda_stream.hpp"
 
 namespace dca {
 namespace linalg {
 namespace util {
 // dca::linalg::util::
 
-class MagmaQueue {
+class MagmaQueue : public linalg::util::CudaStream {
```

Comment:
I think a MagmaQueue has a CudaStream I don't think it is a CudaStream.

Reply:
We could argue if a MagmaQueue is a CudaStream with extra resources. But according to the Liskov Substitution Principle inheritance is appropriate: everywhere you can use a CudaStream you can also use a MagmaQueue.

Reply:
I think its pretty clear a CudaStream is a resource a MagmaQueue owns or has a reference to.  When Magma is build on top of AMD's libraries it clearly won't have a CudaStream or be a cudastream.  I don't like to see inheritance used as a way to get simplified access to class members.

Reply:
AMD HIP tries to follow CUDA. I think there is still "HIP stream"  that works like a cuda stream.

Reply:
you just need to change the implementation of CudaStream (and possibly rename it), this is completely orthogonal from the decision of using composition vs inheritance.
The inheritance has no access to the private members, but it exposes the interface, which I want, and has a well defined downcast that I also need. Of course it could be implemented as composition, but then I need to forward the methods, and use a getter to the stream in the rest of the code. While we don't need runtime polymorphism, I am pretty sure this is a well defined  inheritance, and the added 64 bits of storage for the (unused) vtable pointer is a non issue.